### PR TITLE
feat: add hardhat_setCode

### LINKED ISSUE #149: hardhat_api: Implement `hardhat_setCode` method
## **Description:**
We need to implement the [`hardhat_setCode`](https://github.com/NomicFoundation/hardhat/blob/781161c00a467a1da392d684950eec4d8574b868/packages/hardhat-network-helpers/src/helpers/setCode.ts#L13) method. This function modifies the bytecode stored at a specified account's address.

```rust
// Interface in the `trait`
#[rpc(name = "hardhat_setCode")]
fn set_code(
    &self,
    address: Address, // The address where the given code should be stored
    code: Vec<u8>, // The bytecode to store at the given address
) -> BoxFuture<Result<()>>;

// Actual code in the `impl`
fn set_code(
    &self,
    address: Address,
    code: Vec<u8>,
) -> BoxFuture<Result<()>> {
    // TODO: Solution here
}
```

#### Expected command and response:
Command:
```
curl -X POST http://localhost:8011 \
     -H "Content-Type: application/json" \
     --data '{
       "jsonrpc": "2.0",
       "id": 1,
       "method": "hardhat_setCode",
       "params": ["0x0d2026b3EE6eC71FC6746ADb6311F6d3Ba1C000B", "0xa1a2a3..."]
     }'
```

Response:
```
{"jsonrpc":"2.0", "result":true, "id":1}
```

### **Requirements:**

1. **Implement the Method:**
    - The method should accept `address` of type `Address` and `code` of type `Vec<u8>`.
    - Return an acknowledgment once the bytecode has been set for the given address.

2. **Unit Tests:**
    - Write unit tests ensuring the `hardhat_setCode` method functions correctly.
    - Test with different addresses and bytecode values, including invalid or non-existent ones.
    - Properly handle potential exceptions and errors.

3. **Inline Documentation:**
    - Write inline Rust documentation (`///` comments) explaining the purpose and usage of the `hardhat_setCode` method.
    - Include details about parameters, return values, and any possible exceptions or errors.
    - Update `SUPPORTED_APIS.md` with updated documentation and examples.
    - (Optional) Add request to `test_endpoints.http` for easier manual verification.

### **Task Checklist:**
- [ ] Implement the `hardhat_setCode` method.
- [ ] Write unit tests for the `hardhat_setCode` method.
- [ ] Add inline documentation for the `hardhat_setCode` method.
- [ ] Update documentation for the `hardhat_dropTransaction` endpoint in `SUPPORTED_APIS.md`.

### **Additional Notes:**
If you're an external contributor wanting to work on this issue, please comment below to notify us. If you have questions about the requirements or need any assistance related to the implementation, don't hesitate to ask!

### PR DESCRIPTION
**Pull request by nbaztec**
# What :computer: 
* adds `hardhat_setCode`

# Why :hand:
* Allows developers to set arbitrary smart contract bytecodes for a given address, aiding in development.

# Evidence :camera:
image
image

# Notes :memo:
* Fixes #149

### CODE REVIEW

File: e2e-tests/test/hardhat-apis.test.ts [hunk 1]
```diff
@@ -84,3 +88,34 @@ xdescribe("hardhat_impersonateAccount & hardhat_stopImpersonatingAccount", funct
...
+    const wallet = new Wallet(RichAccounts[0].PrivateKey);
+    const deployer = new Deployer(hre, wallet);
+
+    const greeter = await deployContract(deployer, "Greeter", ["Hi"]);
+    expect(await greeter.greet()).to.eq("Hi");
+
+    // Act
+    const artifact = await deployer.loadArtifact("Return5");
+    const newContractCode = [...ethers.utils.arrayify(artifact.deployedBytecode)];
+    await provider.send("hardhat_setCode", [greeter.address, newContractCode]);
```
Comment (MexicanAce) [inline e2e-tests/test/hardhat-apis.test.ts:104]: nit: Only this line is what's being tested, and thus should be the only line in the `// Act` section. Getting different contract code and loading an artifact is part of test setup/arrangement

File: e2e-tests/test/hardhat-apis.test.ts [hunk 2]
```diff
@@ -84,3 +88,34 @@ xdescribe("hardhat_impersonateAccount & hardhat_stopImpersonatingAccount", funct
     expect(await provider.getBalance(RichAccounts[0].Account)).to.equal(beforeBalance.sub(0.42));
   });
 });
+
+describe("hardhat_setCode", function () {
+  it("Should update code with a different smart contract", async function () {
```
Comment (MexicanAce) [inline e2e-tests/test/hardhat-apis.test.ts:93]: Can we add a test for passing in an address that isn't assigned to a contract/account?

File: e2e-tests/test/hardhat-apis.test.ts [hunk 3]
```diff
@@ -84,3 +88,64 @@ xdescribe("hardhat_impersonateAccount & hardhat_stopImpersonatingAccount", funct
     expect(await provider.getBalance(RichAccounts[0].Account)).to.equal(beforeBalance.sub(0.42));
   });
 });
+
+describe("hardhat_setCode", function () {
+  it.only("Should set code at an address", async function () {
```
Comment (MexicanAce) [inline e2e-tests/test/hardhat-apis.test.ts:93]: Please remove the `.only`. I'll make a ticket to ensure this doesn't happen again and use something like https://stackoverflow.com/a/54015366
```suggestion
  it("Should set code at an address", async function () {
```

File: e2e-tests/test/hardhat-apis.test.ts [hunk 4]
```diff
@@ -84,3 +88,64 @@ xdescribe("hardhat_impersonateAccount & hardhat_stopImpersonatingAccount", funct
...
+        value: "0x0",
+        nonce: "0x1",
+      },
+      "latest",
+    ]);
+    expect(BigNumber.from(result).toNumber()).to.eq(5);
+  });
+});
+
+describe("hardhat_setCode", function () {
```
Comment (MexicanAce) [inline e2e-tests/test/hardhat-apis.test.ts:122]: No need for a separate `describe`, you can have multiple `it(...)` inside of a `describe`

Comment (nbaztec) [reply e2e-tests/test/hardhat-apis.test.ts]: I generally avoid having multiple `it` inside a `describe` as most `describe`s run in isolation but I guess that's not the case here so we can have them together.

**Review by MexicanAce — APPROVED** [review]
LGTM

**Pull request merged**